kuma-cp crashes with empty top level targetRef

### What happened?

Steps to reproduce:

1. start k3d cluster
2. apply policy (I also installed demo and observability but this shouldn't matter here, contact me if cannot reproduce)

```bash
echo 'apiVersion: kuma.io/v1alpha1
kind: MeshTrace
metadata:
  name: trace-all
  namespace: kong-mesh-system
spec:
  default:
    backends:
      - zipkin:
          url: http://jaeger-collector.mesh-observability:9411/api/v2/spans' | kubectl apply -f -
```

Stack trace:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x2517c82]

goroutine 1090 [running]:
github.com/kumahq/kuma/api/common/v1alpha1.(*TargetRef).GetKindEnum(...)
        /home/krzysztof.slonka/go/pkg/mod/github.com/kumahq/kuma@v0.0.0-20221017162450-da91c5000546/api/common/v1alpha1/targetref_helper.go:4
github.com/kumahq/kuma/pkg/plugins/policies/matchers.inboundsSelectedByTargetRef(0x0, 0x41865c0?, 0xc001a6d740?)
        /home/krzysztof.slonka/go/pkg/mod/github.com/kumahq/kuma@v0.0.0-20221017162450-da91c5000546/pkg/plugins/policies/matchers/match.go:118 +0x42
github.com/kumahq/kuma/pkg/plugins/policies/matchers.MatchedPolicies({0x2ec7fd9, 0x9}, 0xc0020c3e48, {0xc0020f0120?, 0xc0020f0150?})
```

--- a/pkg/api-server/inspect_endpoints_test.go
+++ b/pkg/api-server/inspect_endpoints_test.go
@@ -339,6 +339,9 @@ var _ = Describe("Inspect WS", func() {
 						},
 						From: []*policies_api.MeshTrafficPermission_From{
 							{
+								TargetRef: &common_api.TargetRef{
+									Kind: "Mesh",
+								},
 								Default: &policies_api.MeshTrafficPermission_Conf{
 									Action: "ALLOW",
 								},
@@ -1105,6 +1108,9 @@ var _ = Describe("Inspect WS", func() {
 						},
 						From: []*policies_api.MeshTrafficPermission_From{
 							{
+								TargetRef: &common_api.TargetRef{
+									Kind: "Mesh",
+								},
 								Default: &policies_api.MeshTrafficPermission_Conf{
 									Action: "ALLOW",
 								},

--- a/pkg/api-server/testdata/inspect_dataplane.json
+++ b/pkg/api-server/testdata/inspect_dataplane.json
@@ -19,6 +19,9 @@
       },
       "from": [
        {
+        "targetRef": {
+         "kind": "Mesh"
+        },
         "default": {
          "action": "ALLOW"
         }

--- a/pkg/core/xds/rules.go
+++ b/pkg/core/xds/rules.go
@@ -218,10 +218,6 @@ func newConf(t reflect.Type) (proto.Message, error) {
 }
 
 func asSubset(tr *common_api.TargetRef) (Subset, error) {
-	if tr == nil {
-		// syntactic sugar, empty targetRef means targetRef{kind: Mesh}
-		return Subset{}, nil
-	}
 	switch tr.GetKindEnum() {
 	case common_api.TargetRef_Mesh:
 		return Subset{}, nil

--- a/pkg/core/xds/testdata/rules/06.policy.yaml
+++ b/pkg/core/xds/testdata/rules/06.policy.yaml
@@ -4,7 +4,6 @@ name: mt-1
 spec:
   targetRef:
     kind: Mesh
-    name: mesh-1
   default:
     tags:
       - name: team

--- a/pkg/core/xds/testdata/rules/07.policy.yaml
+++ b/pkg/core/xds/testdata/rules/07.policy.yaml
@@ -4,7 +4,6 @@ name: mt-1
 spec:
   targetRef:
     kind: Mesh
-    name: mesh-1
   default:
     tags:
       - name: team

--- a/pkg/plugins/policies/matchers/match.go
+++ b/pkg/plugins/policies/matchers/match.go
@@ -33,8 +33,7 @@ func MatchedPolicies(rType core_model.ResourceType, dpp *core_mesh.DataplaneReso
 			return core_xds.TypedMatchingPolicies{}, errors.Errorf("resource type %v doesn't support TargetRef matching", rType)
 		}
 
-		targetRef := spec.GetTargetRef()
-		selectedInbounds := inboundsSelectedByTargetRef(targetRef, dpp, gateway)
+		selectedInbounds := inboundsSelectedByTargetRef(spec.GetTargetRef(), dpp, gateway)
 		if len(selectedInbounds) == 0 {
 			// DPP is not matched by the policy
 			continue

--- a/pkg/plugins/policies/matchers/testdata/match/dataplanepolicies/01.golden.yaml
+++ b/pkg/plugins/policies/matchers/testdata/match/dataplanepolicies/01.golden.yaml
@@ -9,7 +9,6 @@ items:
         action: ALLOW
       targetRef:
         kind: Mesh
-        name: default
     targetRef:
       kind: Mesh
   type: MeshTrafficPermission
@@ -23,7 +22,6 @@ items:
         action: ALLOW
       targetRef:
         kind: Mesh
-        name: default
     targetRef:
       kind: MeshSubset
       tags:
@@ -39,7 +37,6 @@ items:
         action: ALLOW
       targetRef:
         kind: Mesh
-        name: default
     targetRef:
       kind: MeshService
       name: web
@@ -54,7 +51,6 @@ items:
         action: ALLOW
       targetRef:
         kind: Mesh
-        name: default
     targetRef:
       kind: MeshServiceSubset
       name: web

--- a/pkg/plugins/policies/matchers/testdata/match/dataplanepolicies/01.policies.yaml
+++ b/pkg/plugins/policies/matchers/testdata/match/dataplanepolicies/01.policies.yaml
@@ -11,7 +11,6 @@ spec:
   from:
     - targetRef:
         kind: Mesh
-        name: default
       default:
         action: ALLOW
 ---
@@ -24,7 +23,6 @@ spec:
   from:
     - targetRef:
         kind: Mesh
-        name: default
       default:
         action: ALLOW
 ---
@@ -38,7 +36,6 @@ spec:
   from:
     - targetRef:
         kind: Mesh
-        name: default
       default:
         action: ALLOW
 ---
@@ -53,7 +50,6 @@ spec:
   from:
     - targetRef:
         kind: Mesh
-        name: default
       default:
         action: ALLOW
 ---
@@ -67,7 +63,6 @@ spec:
   from:
     - targetRef:
         kind: Mesh
-        name: default
       default:
         action: ALLOW
 ---
@@ -83,7 +78,6 @@ spec:
   from:
     - targetRef:
         kind: Mesh
-        name: default
       default:
         action: ALLOW
 ---
@@ -99,6 +93,5 @@ spec:
   from:
     - targetRef:
         kind: Mesh
-        name: default
       default:
         action: ALLOW

--- a/pkg/plugins/policies/matchers/testdata/match/dataplanepolicies/02.golden.yaml
+++ b/pkg/plugins/policies/matchers/testdata/match/dataplanepolicies/02.golden.yaml
@@ -9,7 +9,6 @@ items:
         action: ALLOW
       targetRef:
         kind: Mesh
-        name: default
     targetRef:
       kind: Mesh
   type: MeshTrafficPermission
@@ -23,7 +22,6 @@ items:
         action: ALLOW
       targetRef:
         kind: Mesh
-        name: default
     targetRef:
       kind: Mesh
   type: MeshTrafficPermission
@@ -37,7 +35,6 @@ items:
         action: ALLOW
       targetRef:
         kind: Mesh
-        name: default
     targetRef:
       kind: MeshSubset
       tags:
@@ -53,7 +50,6 @@ items:
         action: ALLOW
       targetRef:
         kind: Mesh
-        name: default
     targetRef:
       kind: MeshSubset
       tags:
@@ -69,7 +65,6 @@ items:
         action: ALLOW
       targetRef:
         kind: Mesh
-        name: default
     targetRef:
       kind: MeshService
       name: web
@@ -84,7 +79,6 @@ items:
         action: ALLOW
       targetRef:
         kind: Mesh
-        name: default
     targetRef:
       kind: MeshService
       name: web
@@ -99,7 +93,6 @@ items:
         action: ALLOW
       targetRef:
         kind: Mesh
-        name: default
     targetRef:
       kind: MeshServiceSubset
       name: web
@@ -116,7 +109,6 @@ items:
         action: ALLOW
       targetRef:
         kind: Mesh
-        name: default
     targetRef:
       kind: MeshServiceSubset
       name: web

--- a/pkg/plugins/policies/matchers/testdata/match/dataplanepolicies/02.policies.yaml
+++ b/pkg/plugins/policies/matchers/testdata/match/dataplanepolicies/02.policies.yaml
@@ -11,7 +11,6 @@ spec:
   from:
     - targetRef:
         kind: Mesh
-        name: default
       default:
         action: ALLOW
 ---
@@ -27,7 +26,6 @@ spec:
   from:
     - targetRef:
         kind: Mesh
-        name: default
       default:
         action: ALLOW
 ---
@@ -40,7 +38,6 @@ spec:
   from:
     - targetRef:
         kind: Mesh
-        name: default
       default:
         action: ALLOW
 ---
@@ -53,7 +50,6 @@ spec:
   from:
     - targetRef:
         kind: Mesh
-        name: default
       default:
         action: ALLOW
 ---
@@ -67,7 +63,6 @@ spec:
   from:
     - targetRef:
         kind: Mesh
-        name: default
       default:
         action: ALLOW
 ---
@@ -81,7 +76,6 @@ spec:
   from:
     - targetRef:
         kind: Mesh
-        name: default
       default:
         action: ALLOW
 ---
@@ -96,7 +90,6 @@ spec:
   from:
     - targetRef:
         kind: Mesh
-        name: default
       default:
         action: ALLOW
 ---
@@ -111,6 +104,5 @@ spec:
   from:
     - targetRef:
         kind: Mesh
-        name: default
       default:
         action: ALLOW

--- a/pkg/plugins/policies/matchers/testdata/match/dataplanepolicies/03.policies.yaml
+++ b/pkg/plugins/policies/matchers/testdata/match/dataplanepolicies/03.policies.yaml
@@ -13,6 +13,5 @@ spec:
   from:
     - targetRef:
         kind: Mesh
-        name: default
       default:
         action: ALLOW

--- a/pkg/plugins/policies/matchers/testdata/match/dataplanepolicies/04.policies.yaml
+++ b/pkg/plugins/policies/matchers/testdata/match/dataplanepolicies/04.policies.yaml
@@ -13,6 +13,5 @@ spec:
   from:
     - targetRef:
         kind: Mesh
-        name: default
       default:
         action: ALLOW

--- a/pkg/plugins/policies/matchers/validators/validator.go
+++ b/pkg/plugins/policies/matchers/validators/validator.go
@@ -17,55 +17,39 @@ func ValidateTargetRef(
 ) validators.ValidationError {
 	verr := validators.ValidationError{}
 	if ref == nil {
-		// Implicitly treat nil targetRef as kind Mesh
+		verr.AddViolation("", "you need to set a targetRef")
+		return verr
+	}
+	if ref.Kind == "" {
+		verr.AddViolation("kind", "must be set")
 		return verr
 	}
 	if !contains(opts.SupportedKinds, ref.GetKindEnum()) {
 		verr.AddViolation("kind", "value is not supported")
 	} else {
-		switch ref.GetKindEnum() {
+		refKind := ref.GetKindEnum()
+		switch refKind {
 		case common_proto.TargetRef_Mesh:
-			if len(ref.Tags) != 0 {
-				verr.AddViolation("tags", fmt.Sprintf("could not be set with kind %v", common_proto.TargetRef_Mesh))
+			if ref.Name != "" {
+				verr.AddViolation("name", fmt.Sprintf("using name with kind %v is not yet supported", refKind))
 			}
-			if len(ref.Mesh) != 0 {
-				verr.AddViolation("mesh", fmt.Sprintf("could not be set with kind %v", common_proto.TargetRef_Mesh))
-			}
+			verr.Add(disallowedField("mesh", ref.Mesh, refKind))
+			verr.Add(disallowedField("tags", ref.Tags, refKind))
 		case common_proto.TargetRef_MeshSubset:
-			if len(ref.Mesh) != 0 {
-				verr.AddViolation("mesh", fmt.Sprintf("could not be set with kind %v", common_proto.TargetRef_MeshSubset))
-			}
-			if ref.Tags != nil && len(ref.Tags) == 0 {
-				verr.AddViolation("tags", "cannot be empty")
-			}
+			verr.Add(disallowedField("name", ref.Name, refKind))
+			verr.Add(disallowedField("mesh", ref.Mesh, refKind))
 		case common_proto.TargetRef_MeshService:
-			if len(ref.Tags) != 0 {
-				verr.AddViolation("tags", fmt.Sprintf("could not be set with kind %v", common_proto.TargetRef_MeshService))
-			}
-			if len(ref.Name) == 0 {
-				verr.AddViolation("name", "cannot be empty")
-			}
+			verr.Add(requiredField("name", ref.Name, refKind))
+			verr.Add(disallowedField("mesh", ref.Mesh, refKind))
+			verr.Add(disallowedField("tags", ref.Tags, refKind))
 		case common_proto.TargetRef_MeshServiceSubset:
-			if len(ref.Name) == 0 {
-				verr.AddViolation("name", "cannot be empty")
-			}
-			if ref.Tags != nil && len(ref.Tags) == 0 {
-				verr.AddViolation("tags", "cannot be empty")
-			}
+			verr.Add(requiredField("name", ref.Name, refKind))
+			verr.Add(disallowedField("mesh", ref.Mesh, refKind))
 		case common_proto.TargetRef_MeshGatewayRoute:
-			if len(ref.Name) == 0 {
-				verr.AddViolation("name", "cannot be empty")
-			}
-			if len(ref.Mesh) != 0 {
-				verr.AddViolation("mesh", fmt.Sprintf("could not be set with kind %v", common_proto.TargetRef_MeshGatewayRoute))
-			}
+			verr.Add(requiredField("name", ref.Name, refKind))
+			verr.Add(disallowedField("mesh", ref.Mesh, refKind))
 		case common_proto.TargetRef_MeshHTTPRoute:
-			if len(ref.Name) == 0 {
-				verr.AddViolation("name", "cannot be empty")
-			}
-			if len(ref.Mesh) != 0 {
-				verr.AddViolation("mesh", fmt.Sprintf("could not be set with kind %v", common_proto.TargetRef_MeshHTTPRoute))
-			}
+			verr.AddViolation("kind", fmt.Sprintf("%v is not yet supported", refKind))
 		}
 	}
 	return verr
@@ -76,6 +60,36 @@ func contains(array []common_proto.TargetRef_Kind, item common_proto.TargetRef_K
 		if it == item {
 			return true
 		}
+	}
+	return false
+}
+
+func disallowedField(name string, value interface{}, kind common_proto.TargetRef_Kind) validators.ValidationError {
+	res := validators.ValidationError{}
+	if isSet(value) {
+		res.Violations = append(res.Violations, validators.Violation{
+			Field: name, Message: fmt.Sprintf("cannot be set with kind %v", kind),
+		})
+	}
+	return res
+}
+
+func requiredField(name string, value interface{}, kind common_proto.TargetRef_Kind) validators.ValidationError {
+	res := validators.ValidationError{}
+	if !isSet(value) {
+		res.Violations = append(res.Violations, validators.Violation{
+			Field: name, Message: fmt.Sprintf("must be set with kind %v", kind),
+		})
+	}
+	return res
+}
+
+func isSet(value interface{}) bool {
+	if v, ok := value.(string); ok {
+		return v != ""
+	}
+	if v, ok := value.(map[string]string); ok {
+		return len(v) > 0
 	}
 	return false
 }

--- a/pkg/plugins/policies/matchers/validators/validator_test.go
+++ b/pkg/plugins/policies/matchers/validators/validator_test.go
@@ -13,384 +13,397 @@ import (
 
 var _ = Describe("TargetRef Validator", func() {
 
-	Context("pass validation", func() {
+	type testCase struct {
+		inputYaml string
+		opts      *matcher_validators.ValidateTargetRefOpts
+		// only used for failed testCases
+		expected string
+	}
 
-		type testCase struct {
-			inputYaml string
-			opts      *matcher_validators.ValidateTargetRefOpts
-		}
+	DescribeTable("should pass validation with",
+		func(given testCase) {
+			// given
+			Expect(given.expected).To(BeEmpty())
+			targetRef := &common_proto.TargetRef{}
+			err := util_proto.FromYAML([]byte(given.inputYaml), targetRef)
+			Expect(err).ToNot(HaveOccurred())
 
-		DescribeTable("should pass validation",
-			func(given testCase) {
-				// given
-				targetRef := &common_proto.TargetRef{}
-				err := util_proto.FromYAML([]byte(given.inputYaml), targetRef)
-				Expect(err).ToNot(HaveOccurred())
+			// when
+			validationErr := validators.ValidationError{}
+			validationErr.AddError("targetRef", matcher_validators.ValidateTargetRef(targetRef, given.opts))
 
-				// when
-				validationErr := validators.ValidationError{}
-				validationErr.AddError("targetRef", matcher_validators.ValidateTargetRef(targetRef, given.opts))
-
-				// then
-				Expect(validationErr.OrNil()).To(BeNil())
+			// then
+			Expect(validationErr.OrNil()).To(BeNil())
+		},
+		Entry("Mesh", testCase{
+			inputYaml: `
+kind: Mesh
+`,
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_Mesh,
+				},
 			},
-			Entry("targetRef for Mesh", testCase{
-				inputYaml: `
-kind: Mesh
-`,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_Mesh,
-					},
-				},
-			}),
-			Entry("targetRef for Mesh with name", testCase{
-				inputYaml: `
-kind: Mesh
-name: mesh-1
-`,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_Mesh,
-					},
-				},
-			}),
-			Entry("targetRef for MeshSubset", testCase{
-				inputYaml: `
+		}),
+		Entry("MeshSubset with tags", testCase{
+			inputYaml: `
 kind: MeshSubset
 tags:
   kuma.io/zone: us-east
 `,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshSubset,
-					},
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshSubset,
 				},
-			}),
-			Entry("targetRef for MeshSubset with name", testCase{
-				inputYaml: `
+			},
+		}),
+		Entry("MeshSubset without tags", testCase{
+			inputYaml: `
 kind: MeshSubset
-name: mesh-1
-tags:
-  kuma.io/zone: us-east
 `,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshSubset,
-					},
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshSubset,
 				},
-			}),
-			Entry("targetRef for MeshSubset with name without tags", testCase{
-				inputYaml: `
-kind: MeshSubset
-name: mesh-1
-`,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshSubset,
-					},
-				},
-			}),
-			Entry("targetRef for MeshService", testCase{
-				inputYaml: `
+			},
+		}),
+		Entry("MeshService", testCase{
+			inputYaml: `
 kind: MeshService
 name: backend
 `,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshService,
-					},
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshService,
 				},
-			}),
-			Entry("targetRef for MeshService with mesh", testCase{
-				inputYaml: `
-kind: MeshService
-name: backend
-mesh: mesh-1
-`,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshService,
-					},
-				},
-			}),
-			Entry("targetRef for MeshServiceSubset", testCase{
-				inputYaml: `
+			},
+		}),
+		Entry("MeshServiceSubset", testCase{
+			inputYaml: `
 kind: MeshServiceSubset
 name: backend
 tags:
   version: v1
 `,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshServiceSubset,
-					},
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshServiceSubset,
 				},
-			}),
-			Entry("targetRef for MeshServiceSubset with mesh", testCase{
-				inputYaml: `
+			},
+		}),
+		Entry("MeshServiceSubset without tags", testCase{
+			inputYaml: `
 kind: MeshServiceSubset
 name: backend
-mesh: mesh-1
-tags:
-  version: v1
 `,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshServiceSubset,
-					},
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshServiceSubset,
 				},
-			}),
-			Entry("targetRef for MeshGatewayRoute", testCase{
-				inputYaml: `
+			},
+		}),
+		Entry("MeshServiceSubset with empty tags", testCase{
+			inputYaml: `
+kind: MeshServiceSubset
+name: backend
+tags: {}
+`,
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshServiceSubset,
+				},
+			},
+		}),
+		Entry("MeshGatewayRoute", testCase{
+			inputYaml: `
 kind: MeshGatewayRoute
 name: backend-gateway-route
 `,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshGatewayRoute,
-					},
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshGatewayRoute,
 				},
-			}),
-			Entry("targetRef for MeshHTTPRoute", testCase{
-				inputYaml: `
+			},
+		}),
+	)
+
+	DescribeTable("should return as much individual errors as possible with",
+		func(given testCase) {
+			// given
+			targetRef := &common_proto.TargetRef{}
+			err := util_proto.FromYAML([]byte(given.inputYaml), targetRef)
+			Expect(err).ToNot(HaveOccurred())
+
+			// when
+			validationErr := validators.ValidationError{}
+			validationErr.AddError("targetRef", matcher_validators.ValidateTargetRef(targetRef, given.opts))
+			// and
+			actual, err := yaml.Marshal(validationErr)
+
+			// then
+			Expect(err).ToNot(HaveOccurred())
+			Expect(actual).To(MatchYAML(given.expected))
+		},
+		Entry("empty", testCase{
+			inputYaml: `
+{}
+`,
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_Mesh,
+				},
+			},
+			expected: `
+violations:
+  - field: targetRef.kind
+    message: must be set 
+`,
+		}),
+		Entry("Mesh when it's not supported", testCase{
+			inputYaml: `
+kind: Mesh
+`,
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshSubset,
+				},
+			},
+			expected: `
+violations:
+  - field: targetRef.kind
+    message: value is not supported
+`,
+		}),
+		Entry("Mesh with mesh and tags", testCase{
+			inputYaml: `
+kind: Mesh
+mesh: mesh-1
+tags:
+  tag1: value1
+`,
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_Mesh,
+				},
+			},
+			expected: `
+violations:
+  - field: targetRef.mesh
+    message: cannot be set with kind Mesh
+  - field: targetRef.tags
+    message: cannot be set with kind Mesh
+`,
+		}),
+		Entry("Mesh with name", testCase{
+			inputYaml: `
+kind: Mesh
+name: mesh-1
+`,
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_Mesh,
+				},
+			},
+			expected: `
+violations:
+  - field: targetRef.name
+    message: using name with kind Mesh is not yet supported 
+`,
+		}),
+		Entry("MeshSubset when it's not supported", testCase{
+			inputYaml: `
+kind: MeshSubset
+`,
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_Mesh,
+				},
+			},
+			expected: `
+violations:
+  - field: targetRef.kind
+    message: value is not supported`,
+		}),
+		Entry("MeshSubset with name", testCase{
+			inputYaml: `
+kind: MeshSubset
+name: mesh-1
+`,
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshSubset,
+				},
+			},
+			expected: `
+violations:
+  - field: targetRef.name
+    message: cannot be set with kind MeshSubset`,
+		}),
+		Entry("MeshService when it's not supported", testCase{
+			inputYaml: `
+kind: MeshService
+`,
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshServiceSubset,
+				},
+			},
+			expected: `
+violations:
+  - field: targetRef.kind
+    message: value is not supported`,
+		}),
+		Entry("MeshService with mesh", testCase{
+			inputYaml: `
+kind: MeshService
+name: backend
+mesh: mesh-1
+`,
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshService,
+				},
+			},
+			expected: `
+violations:
+  - field: targetRef.mesh
+    message: cannot be set with kind MeshService
+`,
+		}),
+		Entry("MeshService without name with tags", testCase{
+			inputYaml: `
+kind: MeshService
+tags:
+  tag1: value1
+`,
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshService,
+				},
+			},
+			expected: `
+violations:
+  - field: targetRef.name
+    message: must be set with kind MeshService 
+  - field: targetRef.tags
+    message: cannot be set with kind MeshService
+`,
+		}),
+		Entry("MeshServiceSubset when it's not supported", testCase{
+			inputYaml: `
+kind: MeshServiceSubset
+`,
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshService,
+				},
+			},
+			expected: `
+violations:
+  - field: targetRef.kind
+    message: value is not supported
+`,
+		}),
+		Entry("MeshServiceSubset without name with empty tags", testCase{
+			inputYaml: `
+kind: MeshServiceSubset
+tags: {}
+`,
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshServiceSubset,
+				},
+			},
+			expected: `
+violations:
+  - field: targetRef.name
+    message: must be set with kind MeshServiceSubset
+`,
+		}),
+		Entry("MeshServiceSubset with mesh", testCase{
+			inputYaml: `
+kind: MeshServiceSubset
+name: backend
+mesh: mesh-1
+tags:
+  version: v1
+`,
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshServiceSubset,
+				},
+			},
+			expected: `
+violations:
+  - field: targetRef.mesh
+    message: cannot be set with kind MeshServiceSubset 
+`,
+		}),
+		Entry("MeshGatewayRoute when it's not supported", testCase{
+			inputYaml: `
+kind: MeshGatewayRoute
+`,
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshHTTPRoute,
+				},
+			},
+			expected: `
+violations:
+  - field: targetRef.kind
+    message: value is not supported
+`,
+		}),
+		Entry("MeshGatewayRoute without name with mesh and tags", testCase{
+			inputYaml: `
+kind: MeshGatewayRoute
+mesh: mesh-1
+tags:
+  tag1: value1
+`,
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshGatewayRoute,
+				},
+			},
+			expected: `
+violations:
+  - field: targetRef.name
+    message: must be set with kind MeshGatewayRoute
+  - field: targetRef.mesh
+    message: cannot be set with kind MeshGatewayRoute
+`,
+		}),
+		Entry("MeshHTTPRoute when it's not supported", testCase{
+			inputYaml: `
+kind: MeshHTTPRoute
+`,
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshGatewayRoute,
+				},
+			},
+			expected: `
+violations:
+  - field: targetRef.kind
+    message: value is not supported
+`,
+		}),
+		Entry("MeshHTTPRoute as it's not currently supported", testCase{
+			inputYaml: `
 kind: MeshHTTPRoute
 name: backend-http-route
 `,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshHTTPRoute,
-					},
+			opts: &matcher_validators.ValidateTargetRefOpts{
+				SupportedKinds: []common_proto.TargetRef_Kind{
+					common_proto.TargetRef_MeshHTTPRoute,
 				},
-			}),
-		)
-	})
-
-	Context("fail validation", func() {
-
-		type testCase struct {
-			inputYaml string
-			opts      *matcher_validators.ValidateTargetRefOpts
-			expected  string
-		}
-
-		DescribeTable("should validate all fields and return as much individual errors as possible",
-			func(given testCase) {
-				// given
-				targetRef := &common_proto.TargetRef{}
-				err := util_proto.FromYAML([]byte(given.inputYaml), targetRef)
-				Expect(err).ToNot(HaveOccurred())
-
-				// when
-				validationErr := validators.ValidationError{}
-				validationErr.AddError("targetRef", matcher_validators.ValidateTargetRef(targetRef, given.opts))
-				// and
-				actual, err := yaml.Marshal(validationErr)
-
-				// then
-				Expect(err).ToNot(HaveOccurred())
-				Expect(actual).To(MatchYAML(given.expected))
 			},
-			Entry("targetRef for Mesh when it's not supported", testCase{
-				inputYaml: `
-kind: Mesh
-`,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshSubset,
-					},
-				},
-				expected: `
+			expected: `
 violations:
   - field: targetRef.kind
-    message: value is not supported
+    message: MeshHTTPRoute is not yet supported 
 `,
-			}),
-			Entry("targetRef for Mesh with mesh and tags", testCase{
-				inputYaml: `
-kind: Mesh
-mesh: mesh-1
-tags:
-  tag1: value1
-`,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_Mesh,
-					},
-				},
-				expected: `
-violations:
-  - field: targetRef.tags
-    message: could not be set with kind Mesh
-  - field: targetRef.mesh
-    message: could not be set with kind Mesh
-`,
-			}),
-			Entry("targetRef for MeshSubset when it's not supported", testCase{
-				inputYaml: `
-kind: MeshSubset
-`,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_Mesh,
-					},
-				},
-				expected: `
-violations:
-  - field: targetRef.kind
-    message: value is not supported`,
-			}),
-			Entry("targetRef for MeshSubset with empty tags", testCase{
-				inputYaml: `
-kind: MeshSubset
-tags: {}
-`,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshSubset,
-					},
-				},
-				expected: `
-violations:
-  - field: targetRef.tags
-    message: cannot be empty`,
-			}),
-			Entry("targetRef for MeshService when it's not supported", testCase{
-				inputYaml: `
-kind: MeshService
-`,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshServiceSubset,
-					},
-				},
-				expected: `
-violations:
-  - field: targetRef.kind
-    message: value is not supported`,
-			}),
-			Entry("targetRef for MeshService without name with tags", testCase{
-				inputYaml: `
-kind: MeshService
-tags:
-  tag1: value1
-`,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshService,
-					},
-				},
-				expected: `
-violations:
-  - field: targetRef.tags
-    message: could not be set with kind MeshService
-  - field: targetRef.name
-    message: cannot be empty
-`,
-			}),
-			Entry("targetRef for MeshServiceSubset when it's not supported", testCase{
-				inputYaml: `
-kind: MeshServiceSubset
-`,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshService,
-					},
-				},
-				expected: `
-violations:
-  - field: targetRef.kind
-    message: value is not supported
-`,
-			}),
-			Entry("targetRef for MeshServiceSubset without name with empty tags", testCase{
-				inputYaml: `
-kind: MeshServiceSubset
-tags: {}
-`,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshServiceSubset,
-					},
-				},
-				expected: `
-violations:
-  - field: targetRef.name
-    message: cannot be empty
-  - field: targetRef.tags
-    message: cannot be empty
-`,
-			}),
-			Entry("targetRef for MeshGatewayRoute when it's not supported", testCase{
-				inputYaml: `
-kind: MeshGatewayRoute
-`,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshHTTPRoute,
-					},
-				},
-				expected: `
-violations:
-  - field: targetRef.kind
-    message: value is not supported
-`,
-			}),
-			Entry("targetRef for MeshGatewayRoute without name with mesh and tags", testCase{
-				inputYaml: `
-kind: MeshGatewayRoute
-mesh: mesh-1
-tags:
-  tag1: value1
-`,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshGatewayRoute,
-					},
-				},
-				expected: `
-violations:
-  - field: targetRef.name
-    message: cannot be empty
-  - field: targetRef.mesh
-    message: could not be set with kind MeshGatewayRoute
-`,
-			}),
-			Entry("targetRef for MeshHTTPRoute when it's not supported", testCase{
-				inputYaml: `
-kind: MeshHTTPRoute
-`,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshGatewayRoute,
-					},
-				},
-				expected: `
-violations:
-  - field: targetRef.kind
-    message: value is not supported
-`,
-			}),
-			Entry("targetRef for MeshHTTPRoute without name with mesh and tags", testCase{
-				inputYaml: `
-kind: MeshHTTPRoute
-mesh: mesh-1
-tags:
-  tag1: value1
-`,
-				opts: &matcher_validators.ValidateTargetRefOpts{
-					SupportedKinds: []common_proto.TargetRef_Kind{
-						common_proto.TargetRef_MeshHTTPRoute,
-					},
-				},
-				expected: `
-violations:
-  - field: targetRef.name
-    message: cannot be empty
-  - field: targetRef.mesh
-    message: could not be set with kind MeshHTTPRoute
-`,
-			}),
-		)
-	})
+		}),
+	)
 })

--- a/pkg/plugins/policies/meshaccesslog/api/v1alpha1/validator_test.go
+++ b/pkg/plugins/policies/meshaccesslog/api/v1alpha1/validator_test.go
@@ -32,7 +32,6 @@ targetRef:
 from:
   - targetRef:
       kind: Mesh
-      name: default
     default:
       backends:
         - tcp:
@@ -59,7 +58,6 @@ targetRef:
 from:
   - targetRef:
       kind: Mesh
-      name: default
     default:
       backends:
         - file:
@@ -92,7 +90,6 @@ from:
 				inputYaml: `
 targetRef:
   kind: Mesh
-  name: default
 `,
 				expected: `
 violations:
@@ -107,7 +104,6 @@ targetRef:
 from:
   - targetRef:
       kind: Mesh
-      name: default
     default:
       backends:
         - file:
@@ -127,7 +123,6 @@ targetRef:
 from:
   - targetRef:
       kind: Mesh
-      name: default
     default:
       backends:
         - file:
@@ -148,7 +143,6 @@ targetRef:
 from:
   - targetRef:
       kind: Mesh
-      name: default
     default:
       backends:
         - file:
@@ -170,7 +164,6 @@ targetRef:
 from:
   - targetRef:
       kind: Mesh
-      name: default
     default:
       backends:
         - file:
@@ -192,7 +185,6 @@ targetRef:
 from:
   - targetRef:
       kind: Mesh
-      name: default
     default:
       backends:
         - file:
@@ -215,7 +207,6 @@ targetRef:
 from:
   - targetRef:
       kind: Mesh
-      name: default
     default:
       backends:
         - tcp:
@@ -239,7 +230,6 @@ targetRef:
 from:
   - targetRef:
       kind: Mesh
-      name: default
     default:
       backends:
         - tcp:
@@ -267,7 +257,6 @@ targetRef:
 to:
   - targetRef:
       kind: Mesh
-      name: default
     default:
       backends:
         - file:
@@ -288,7 +277,6 @@ targetRef:
 to:
   - targetRef:
       kind: Mesh
-      name: default
     default:
       backends:
         - file:
@@ -298,6 +286,8 @@ to:
 `,
 				expected: `
 violations:
+- field: spec.targetRef.kind
+  message: MeshHTTPRoute is not yet supported
 - field: spec.to
   message: 'cannot use "to" when "targetRef" is "MeshHTTPRoute" - "to" always goes to the application'`,
 			}),
@@ -305,11 +295,9 @@ violations:
 				inputYaml: `
 targetRef:
   kind: Mesh
-  name: default
 to:
   - targetRef:
       kind: Mesh
-      name: default
 `,
 				expected: `
 violations:
@@ -320,11 +308,9 @@ violations:
 				inputYaml: `
 targetRef:
   kind: Mesh
-  name: default
 from:
   - targetRef:
       kind: Mesh
-      name: default
 `,
 				expected: `
 violations:
@@ -339,7 +325,6 @@ targetRef:
 from:
   - targetRef:
       kind: Mesh
-      name: default
     default:
       backends:
         - tcp:

--- a/pkg/plugins/policies/meshtrafficpermission/api/v1alpha1/validator_test.go
+++ b/pkg/plugins/policies/meshtrafficpermission/api/v1alpha1/validator_test.go
@@ -53,18 +53,15 @@ from:
     default:
       action: DENY
 `),
-			Entry("allow empty targetRefs", `
-from:
-  - default:
-      action: DENY
-`),
 			Entry("allow MeshSubset at top-level targetRef", `
 targetRef:
   kind: MeshSubset
   tags:
     env: prod
 from:
-  - default:
+  - targetRef:
+      kind: Mesh
+    default:
       action: DENY
 `),
 			Entry("allow MeshService at top-level targetRef", `
@@ -72,7 +69,9 @@ targetRef:
   kind: MeshService
   name: backend
 from:
-  - default:
+  - targetRef:
+      kind: Mesh
+    default:
       action: DENY
 `),
 			Entry("allow MeshServiceSubset at top-level targetRef", `
@@ -82,7 +81,9 @@ targetRef:
   tags:
     version: v2
 from:
-  - default:
+  - targetRef:
+      kind: Mesh
+    default:
       action: DENY
 `),
 			Entry("allow MeshGatewayRoute at top-level targetRef", `
@@ -90,15 +91,9 @@ targetRef:
   kind: MeshGatewayRoute
   name: backend-gateway-route
 from:
-  - default:
-      action: DENY
-`),
-			Entry("allow MeshHTTPRoute at top-level targetRef", `
-targetRef:
-  kind: MeshHTTPRoute
-  name: backend-http-route
-from:
-  - default:
+  - targetRef:
+      kind: Mesh
+    default:
       action: DENY
 `),
 		)
@@ -124,6 +119,22 @@ from:
 				// then
 				Expect(actual).To(MatchYAML(given.expected))
 			},
+			Entry("disallow MeshHTTPRoute at top-level targetRef", testCase{
+				inputYaml: `
+targetRef:
+  kind: MeshHTTPRoute
+  name: backend-http-route
+from:
+  - targetRef:
+      kind: Mesh
+    default:
+      action: DENY
+`,
+				expected: `
+violations:
+  - field: spec.targetRef.kind
+    message: MeshHTTPRoute is not yet supported`,
+			}),
 			Entry("empty 'from' array", testCase{
 				inputYaml: `
 targetRef:
@@ -175,6 +186,8 @@ violations:
 			}),
 			Entry("default is nil", testCase{
 				inputYaml: `
+targetRef:
+  kind: Mesh
 from:
   - targetRef:
       kind: Mesh

--- a/test/e2e_env/kubernetes/observability/meshtrace.go
+++ b/test/e2e_env/kubernetes/observability/meshtrace.go
@@ -24,7 +24,6 @@ metadata:
 spec:
   targetRef:
     kind: Mesh
-    name: %s
   default:
     tags:
       - name: team
@@ -32,7 +31,7 @@ spec:
     backends:
       - zipkin:
           url: %s
-`, meshName, meshName, url)
+`, meshName, url)
 }
 
 func PluginTest() {

--- a/test/e2e_env/universal/observability/meshtrace.go
+++ b/test/e2e_env/universal/observability/meshtrace.go
@@ -19,7 +19,6 @@ mesh: %s
 spec:
   targetRef:
     kind: Mesh
-    name: %s
   default:
     tags:
       - name: team
@@ -27,7 +26,7 @@ spec:
     backends:
       - zipkin:
           url: %s
-`, meshName, meshName, url)
+`, meshName, url)
 }
 
 func PluginTest() {

